packages/contract/src/builder.ts

```diff
@@ -678,3 +678,422 @@ export function defineContract<TContract extends ContractDefinition>(
+ *
+ * @param exchange - The exchange to publish to (direct or topic type)
+ * @param queue - The queue that will receive the messages
+ * @param message - The message definition (schema and metadata)
+ * @param options - Binding configuration (routingKey is required)
+ * @param options.routingKey - The routing key for message routing
+ * @returns A publisher-first result with publisher, binding, and consumer factory
+ *
+ * @example
+ * ```typescript
+ * import { z } from 'zod';
+ *
+ * const ordersExchange = defineExchange('orders', 'topic', { durable: true });
+ * const orderQueue = defineQueue('order-processing', { durable: true });
+ * const orderMessage = defineMessage(
+ *   z.object({
+ *     orderId: z.string(),
+ *     amount: z.number(),
+ *   })
+ * );
+ *
+ * // Create publisher-first relationship with routing key
+ * const orderPublisherFirst = definePublisherFirst(
+ *   ordersExchange,
+ *   orderQueue,
+ *   orderMessage,
+ *   { routingKey: 'order.created' }
+ * );
+ *
+ * // Use in contract - routing key is consistent across publisher and binding
+ * const contract = defineContract({
+ *   exchanges: { orders: ordersExchange },
+ *   queues: { orderQueue },
+ *   bindings: { orderBinding: orderPublisherFirst.binding },
+ *   publishers: { orderCreated: orderPublisherFirst.publisher },
+ *   consumers: { processOrder: orderPublisherFirst.createConsumer() },
+ * });
+ * ```
+ */
+export function definePublisherFirst<TMessage extends MessageDefinition>(
+  exchange: DirectExchangeDefinition | TopicExchangeDefinition,
+  queue: QueueDefinition,
+  message: TMessage,
+  options: Omit<
+    Extract<
+      QueueBindingDefinition,
+      { exchange: DirectExchangeDefinition | TopicExchangeDefinition }
+    >,
+    "type" | "queue" | "exchange"
+  >,
+): PublisherFirstResult<
+  TMessage,
+  Extract<
+    PublisherDefinition<TMessage>,
+    { exchange: DirectExchangeDefinition | TopicExchangeDefinition }
+  >,
+  Extract<QueueBindingDefinition, { exchange: DirectExchangeDefinition | TopicExchangeDefinition }>
+>;
+
+/**
+ * Implementation of definePublisherFirst.
+ * @internal
+ */
+export function definePublisherFirst<TMessage extends MessageDefinition>(
+  exchange: ExchangeDefinition,
+  queue: QueueDefinition,
+  message: TMessage,
+  options?: {
+    routingKey?: string;
+    arguments?: Record<string, unknown>;
+  },
+): PublisherFirstResult<TMessage, PublisherDefinition<TMessage>, QueueBindingDefinition> {
+  // Create the binding and publisher using helper functions
+  const binding = createQueueBindingForExchange(queue, exchange, options);
+  const publisher = createPublisherForExchange(exchange, message, options);
+
+  // Factory function to create a consumer with the same message type
+  const createConsumer = (): ConsumerDefinition<TMessage> => {
+    return defineConsumer(queue, message);
+  };
+
+  return {
+    publisher,
+    binding,
+    createConsumer,
+  };
+}
```

Comment:
```suggestion
export function definePublisherFirst<TMessage extends MessageDefinition>(
  exchange: ExchangeDefinition,
  message: TMessage,
  options?: {
    routingKey?: string;
    arguments?: Record<string, unknown>;
  },
): PublisherFirstResult<TMessage, PublisherDefinition<TMessage>, QueueBindingDefinition> {
  // Create the binding and publisher using helper functions
  const publisher = createPublisherForExchange(exchange, message, options);

  // Factory function to create a consumer with the same message type
  const createConsumer = (queue: QueueDefinition): ConsumerDefinition<TMessage> => {
    const binding = createQueueBindingForExchange(queue, exchange, options);
    return defineConsumer(queue, message);
  };

  return {
    publisher,
    binding,
    createConsumer,
  };
}
```

publishers to exchange do not have to know queues behind the exchange. my suggestion does seem ok to you ?